Chat: Scope history to user account

philipp-spiess opened the pull request:
Closes #2261

This PR scopes the chat history storage to the user account (we use the email + server URL as the key for this). That way, when switching accounts, you will not see chat histories from other accounts (potentially showing work stuff in your free time).

There's one caveat that I have depicted in the test plan. If you keep a window open while you switch accounts and _send a new message to it_, we will _fork_ the history, so it will appear in your other account with the updated messages but kept as-is in your other account. I think this behavior makes sense though and we should not try to optimize for it (i.e I don't think we should close chat windows, for example, when you switch accounts).

⚠️ I plan to land this after cutting the 1.0.2 patch release and deploy it onto the insider build right away to test this some more.

## Test plan

- Log in to a user from dotcom
- Chat!
- Switch to another account, e.g. S2
- More chatting!
- ...repeat


https://github.com/sourcegraph/cody/assets/458591/a42f1a4b-da0c-4c4e-839a-c3717d4c8bbb




<!-- Required. See https://docs.sourcegraph.com/dev/background-information/testing_principles. -->

toolmantim commented:
Too risky for GA? I imagine it's a deep enough change it would need a bunch of thorough manual QA.

philipp-spiess commented:
@toolmantim yeah, something else came up anyways 😬  This PR did not get ready in time anyways

valerybugakov reviewed (commented):
> I don't think we should close chat windows, for example, when you switch accounts

I would expect that because it's the learned behavior from all the chat products. If I switch accounts on Facebook or the ChatGPT website, I won't be able to see currently opened chats. Curious about the rationale here. 

> If you keep a window open while you switch accounts and send a new message to it, we will fork the history

I bumped into this scenario during local testing. I have multiple chat windows opened, and after I switched an account, all of the chat windows are still there. Then, if I reload the VS Code instance, all the chat windows are there without the messages _but_ with titles from the previous account. It does not feel right. I would expect Cody to wipe all previous chats to keep my data secure and private if I log out.

valerybugakov commented on vscode/src/services/LocalStorageProvider.ts:
```diff
@@ -156,3 +198,11 @@ export class LocalStorage {
  */
 export const localStorage = new LocalStorage()
+
+function getKeyForAuthStatus(authStatus: AuthStatus): `${string}-${string}` {
```
Would it be nice to extract ``${string}-${string}`` into a named type?

philipp-spiess commented on vscode/src/services/LocalStorageProvider.ts:
```diff
@@ -67,34 +69,74 @@ export class LocalStorage {
+        // not scoped to individual user accounts to be scoped instead.
+        if (history && !isChatHistoryV2(history)) {
+            history = {
+                [key]: history as UserLocalHistory,
```
Need to make sure `chat` and `user` is not deleted here to not break when downgrading, see https://sourcegraph.slack.com/archives/C05AGQYD528/p1702954449990249

toolmantim commented:
Yeah, to echo @valerybugakov's comment… I would have expected it to close chats when logging out. Making copies of chats into new accounts seems problematic and confusing. Curious to hear why.

philipp-spiess commented:
No good reason, it was just "the way it worked" and I didn't find it too weird. But agree with both of you, when we log out we should close all open cody tabs. Consider it done

philipp-spiess commented:
@valerybugakov @toolmantim I changed the behavior so that chat panels clear when the authenticated user changes. This indeeds feels much better. I attached a video as the test plan. 

@valerybugakov I have also fixed the issue where when you test this PR you can no longer downgrade 👍

valerybugakov commented on vscode/src/services/LocalStorageProvider.ts:
```diff
@@ -67,34 +71,80 @@ export class LocalStorage {
+
+        const key = getKeyForAuthStatus(authStatus)
+
+        // For backwards compatibility, we upgrade the local storage key from the old layout that is
+        // not scoped to individual user accounts to be scoped instead.
```
💜

valerybugakov commented on vscode/src/services/LocalStorageProvider.ts:
```diff
@@ -156,3 +206,11 @@ export class LocalStorage {
+    return `${authStatus.endpoint}-${authStatus.primaryEmail}`
+}
+
+function isChatHistoryV2(history: { [key: ChatHistoryKey]: UserLocalHistory } | UserLocalHistory): boolean {
+    return !!Object.keys(history).find(k => k.includes('-'))
+}
```
Let's add a comment explaining what V2 means for chat history or change the function name to describe that. Looking at this code from the PR diff, it's obvious but might need to be clarified for folks reading the code later in their editors.

valerybugakov commented on vscode/src/services/LocalStorageProvider.ts:
```diff
@@ -67,34 +71,80 @@ export class LocalStorage {
+
+            // We use a raw write here to ensure we do not _append_ a key but actually replace
+            // existing `chat` and `input` keys.
+            void this.storage.update(this.KEY_LOCAL_HISTORY, history)
```
Do we want to `try/catch` or `await` here, similar to how it's done in `setChatHistory`?

philipp-spiess replied:
That function is non-blocking right now and I wanted to avoid changing the API (since it is called in some constructors AFAIK). We definitely don't want to silence the error though but I think having it write asynchronously is fine. The worst part we call the same function again and the state is not updated yet but that would simply call the same update again and not destruct anything (provided writes are atomic but if they are not it would be v strange I guess?).

valerybugakov commented on vscode/src/services/LocalStorageProvider.ts:
```diff
@@ -67,34 +71,80 @@ export class LocalStorage {
+        if (!Object.hasOwn(history, key)) {
+            return { chat: {}, input: [] }
+        }
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore: TS seems to be consufed about using `key` as the key here
+        return history[key] as UserLocalHistory
```
This can be fixed by adding type-casting the backward compatible history object and converting `isChatHistoryV2` to a type-guard:

```ts
type AccountBasedChatHistory = {
  [key: ChatHistoryKey]: UserLocalHistory
  
  // For backward compatibility, we do not want to delete the `chat` and `input` keys.
  // As otherwise, downgrading to a prior version would completely block the startup
  // as the client would throw.
  chat: UserLocalHistory
  input: []
}

public getChatHistory(authStatus: AuthStatus): UserLocalHistory {
  history = {
      [key]: history,
      chat: {},
      input: [],
  } as AccountBasedChatHistory

  return history[key] 
}

function isChatHistoryV2(history: AccountBasedChatHistory | UserLocalHistory): history is AccountBasedChatHistory {
    return !!Object.keys(history).find(k => k.includes('-'))
}

philipp-spiess replied:
<img width="1374" alt="Screenshot 2023-12-21 at 14 40 06" src="https://github.com/sourcegraph/cody/assets/458591/2676ddc8-72aa-40b6-b6e7-a963e16d6ec7">
 
 😢

philipp-spiess replied:
```
            history = {
                ...{ [key]: history },
                ...{
                    chat: {},
                    input: [],
                },
            } satisfies AccountKeyedChatHistory
```

this works lol but this does not?


```
            history = {
                [key]: history,
                chat: {},
                input: [],
            } satisfies AccountKeyedChatHistory
```

TS never fails to surprise

philipp-spiess replied:
<img width="1431" alt="Screenshot 2023-12-21 at 14 44 27" src="https://github.com/sourcegraph/cody/assets/458591/caa9eef3-7ce4-45dd-b33e-3b2a4b988efa">
Doesn't help with the initial issue either though

valerybugakov replied:
Is my TS server stale or 🤔

philipp-spiess replied:
@valerybugakov I applied the file from the gist and 😬  

```
vscode/src/services/LocalStorageProvider.ts:95:23 - error TS2352: Conversion of type '{ [x: string]: [] | UserLocalHistory | {}; chat: {}; input: []; }' to type 'AccountBasedChatHistory' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Types of property 'chat' are incompatible.
    Type '{}' is missing the following properties from type 'UserLocalHistory': chat, input

 95             history = {
                          ~
 96                 [key]: history,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
 98                 input: [],
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
 99             } as AccountBasedChatHistory
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error.

 ELIFECYCLE  Command failed with exit code 2.
```

valerybugakov replied:
🤷‍♂️

valerybugakov reviewed (approved):
Thanks for addressing all the issues. It works like a charm now!
